psl-core/src/main/java/org/linqs/psl/evaluation/statistics/RankingEvaluator.java

```diff
@@ -289,34 +292,241 @@ public double auroc() {
+                    totalPositives.put(queryId, (totalPositives.get(queryId) + 1));
+                } else {
+                    totalPositives.put(queryId, 1);
+                }
+            }
+        }
+
+//        System.out.println("total positives" + totalPositives.toString());
+//        System.out.println(predicted.toString());
+
+        // find the average precision for each query. Note that the predictions are already sorted
+        // hashtables to hold the true positives seen so far, the position, the running sum of precisions, and
+        // the average precision for each query
+        Map<String, Integer> TPSeen = new Hashtable<>();
+        Map<String, Integer> Position = new Hashtable<>();
+        Map<String, Double> SumP = new Hashtable<>();
+        Map<String, Double> AveP = new Hashtable<>();
+
+        for (GroundAtom atom : predicted) {
+            Boolean label = getTruthLabel(atom);
+            String queryId = atom.getArguments()[0].toString();
+
+            if (label == null) {
+                continue;
+            }
+
+            // always update query position
+            if (Position.containsKey(queryId)) {
+                Position.put(queryId, (Position.get(queryId) + 1));
+            } else {
+                Position.put(queryId, 1);
+            }
+
+            // only update the TPSeen count and the running precision sum if the document is relevant
+            if (label) {
+                if (TPSeen.containsKey(queryId)) {
+                    TPSeen.put(queryId, (TPSeen.get(queryId) + 1));
+                } else {
+                    TPSeen.put(queryId, 1);
+                }
+
+                if (SumP.containsKey(queryId)) {
+                    SumP.put(queryId, (SumP.get(queryId) + ((double) TPSeen.get(queryId) / Position.get(queryId))));
+                } else {
+                    SumP.put(queryId, 1.0);
+                }
+            }
+        }
+
+//        System.out.println("TP seen" + TPSeen.toString());
+//        System.out.println("Position" + Position.toString());
+//        System.out.println("SumP" + SumP.toString());
+
+        // Calculate the average precision for each query
+        for (Map.Entry<String, Double> entry : SumP.entrySet()) {
+            AveP.put(entry.getKey(), (entry.getValue() / totalPositives.get(entry.getKey())));
+        }
+
+//        System.out.println("AveP" + AveP.toString());
+
+        // Calculate the mean average precision over all the queries
+        double MAP = 0.0;
+        int NQ = 0;
+
+        for (Map.Entry<String, Double> entry : AveP.entrySet()) {
+            MAP = MAP + entry.getValue();
+            NQ = NQ + 1;
+        }
+        MAP = MAP / (double)NQ;
+
+        return MAP;
+    }
+
+    public double discountedCumulativeGain() {
+        Map<String, Double> DCG = getDCGTable();
+
+        // Calculate the mean dcg over each query
+        double meanDCG = 0.0;
+        int NQ = 0;
```

Comment:
What does NQ stand for? "Num Queries"?
Either way, use more meaningful names.
Favor maintainability and readability over terseness.